File: 240507/괄호 문자열의 적합성 판단/parentheses-string.java
import java.util.*;

public class Main {
    static Scanner sc = new Scanner(System.in);
    static char[] str = sc.next().toCharArray();
    static Stack stack = new Stack<Character>();

    public static void main(String[] args) {
        if(validBraces()) {
            System.out.println("Yes");
        } else {
            System.out.println("No");
        }
    }

    static boolean validBraces() {
        for(char item: str) {
            if(item == '(') {
                stack.push(item);
            } else if (item == ')'){
                if(stack.isEmpty()) {
                    return false;
                }
                stack.pop();
            }
        }

        if(!stack.isEmpty()) {
            return false;
        }
        return true;
    }
}

File: 240507/괄호 문자열의 적합성 판단/parentheses-string.kt
import java.util.*;

val sc = Scanner(System.`in`)
val stack = Stack<Char>()

fun main() {
    val str = sc.next()

    if (validBrace(str)) {
        println("Yes")
    } else {
        println("No")
    }
}

fun validBrace(str: String): Boolean {
    for(item in str) {
        when (item) {
            '(' -> {
                stack.push(item)
            }
            ')' -> {
                if(stack.isEmpty()) {
                    return false
                }
                stack.pop()
            }
        }
    }
    if(!stack.isEmpty()) { 
        return false
    }
    return true
}